markov-analysis.py

```diff
@@ -0,0 +1,220 @@
+    prefix map.
+
+    # TODO: This looks messy, is there a better way than changing so many types?
+
+    prefix_map: dict of tuples to lists
+    prefix_length: int, the length (in words) of the sequence of prefixes to use
+    m: int, the length (in words) of the list of random words
+
+    dict, int, int -> list
+    '''
+    random_text = list(random.choice(list(prefix_map.keys())))
+
+    while (len(random_text) < text_length):
+        prefix = tuple(random_text[(len(random_text)-prefix_length):len(random_text)])
+        random_text.append(random.choice(prefix_map[prefix]))
+
+    return random_text
+
+def prefix_map(word_list, n):
+    '''
+    Takes a list of words and returns a dict mapping sequences of n words (which
+    we call the prefixes) to a list of all next words (which we call the 
+    suffixes).
+
+    word_list: list
+    n: int, the number of prefixes to consider
+    
+    list, int -> dict
+    '''
+    ''' TODO: This doesn't do quite what I want but can't figure out why
+        prefix_map = {(''):{word_list[0]}} # First word follows nothing
+        if n > 1: # Map starting prefix sequences of length <n
+            for i in range(1, n):
+                prefix_seq = tuple(word_list[1:i])
+                prefix_map[prefix_seq] = {word_list[i]}
+    '''
+    prefix_map = {}
+    # Maps prefix n-tuples to suffix list
+    for prefix_start in range(0, len(word_list)-n):
+        prefix_seq = tuple(word_list[prefix_start:prefix_start+n]) 
+        suffix = word_list[prefix_start+n]
+        prefix_map[prefix_seq] = prefix_map.get(prefix_seq, []) + [suffix]
+    
+    return prefix_map
+
+# WARNING: skip_header=True only known to work with emma.txt
+def process_file(filename, skip_header):
+    """
+    (Taken from chapter 13 of ThinkPython, modified to return list)
+    Makes a list that contains the words from a file.
+
+    filename: string
+    skip_header: boolean, whether to skip the Gutenberg header
+   
+    returns: list of all words in the file in order of appearance
+    """
+    word_list = []
+    fp = open(filename)
+
+    if skip_header:
+        skip_gutenberg_header(fp)
+
+    for line in fp:
+        process_line(line, word_list)
+
+    return word_list
+
+# WARNING: This is only known to work with emma since Project Gutenberg books 
+#          do NOT have a standard header format.
+def skip_gutenberg_header(fp):
+    """
+    (Taken from chapter 13 of ThinkPython)
+    Reads from fp until it finds the line that ends the header.
+
+    TODO: Project Gutenberg books do not have a standard header. Figure out how 
+          to skip the header in general.
+    TODO: Some Project Gutenberg books have additional material after the book.
+          Figure out how to skip that.
+
+    fp: open file object
+    """
+    for line in fp:
+        if line.startswith('*END*THE SMALL PRINT!'):
+            break
+
+def process_line(line, word_list):
```

Comment:
I know this is part of the exercise, but in practice you'd probably want to use a proper tokenizer here.

Reply:
there's a surprising number of edge cases.